src/mpl/include/mpl_gpu.h

```diff
@@ -108,6 +116,10 @@ int MPL_gpu_init_device_mappings(int max_devid, int max_subdev_id);
 int MPL_gpu_fast_memcpy(void *src, MPL_pointer_attr_t * src_attr, void *dest,
                         MPL_pointer_attr_t * dest_attr, size_t size);
 
+int MPL_gpu_imemcpy(void *dest_ptr, void *src_ptr, size_t size, int dev,
+                    MPL_gpu_engine_type_t engine_type, MPL_gpu_request * req, bool commit);
+int MPL_gpu_test(MPL_gpu_request * req, int *completed);
```

Comment:
So this operation is synchronous with the CPU once complete?

Reply:
Sorry I'm not sure I understand the question

Reply:
I'm thinking in terms of what I know from `cudaMemcpyAsync`. Since that takes a stream argument, you can queue other work behind on the same stream it that relies on completion. Since we have no stream here, I'm just confirming that the copy is synchronous with the host/CPU.

Reply:
I see. `MPL_gpu_request` is defined by the backend, so if CUDA would need to include a stream, that would be possible. In the case of ze, in `MPL_gpu_imemcpy` we attach an event to the copy and include that event in the request: 

```
        ret =
            zeCommandListAppendMemoryCopy(cmdList, dest_ptr, src_ptr, size, event,
                                          device_state->prev_event ? 1 : 0,
                                          device_state->
                                          prev_event ? &device_state->prev_event : NULL);
        ZE_ERR_CHECK(ret);
        req->event = event;
```
(https://github.com/pmodels/mpich/pull/6569/files#diff-5663c21994fdf0253354edc7a8e5b2d26ee93c7848a69fa72d6b4eabd69c9da0R1655-R1661)

then simply query the event in `MPL_gpu_test`:

```
    ret = zeEventQueryStatus(req->event);
```
(https://github.com/pmodels/mpich/pull/6569/files#diff-5663c21994fdf0253354edc7a8e5b2d26ee93c7848a69fa72d6b4eabd69c9da0R1702)

So when `MPL_gpu_test` returns `*completed == 1`, the copy should be completed with respect to the host. I could imagine cuda using the same method (calling `cudaEventRecord` after calling the `cudaMemcpyAsync` and querying the event in `MPL_gpu_test`)

Reply:
OK. I think in the future we want to have the ability to queue things on an explicit stream, but this is probably fine as a starting point. Resolving this convo.